docs/source/grammar.rst

```diff
@@ -0,0 +1,242 @@
+dictionary keys (see the definition of ``dictKey`` in the `grammar <https://github.com/omry/omegaconf/blob/master/omegaconf/grammar/OmegaConfGrammarParser.g4>`_).
+
+Finally, a ``primitive`` is everything else that is allowed, including in particular (see the `full grammar <https://github.com/omry/omegaconf/blob/master/omegaconf/grammar/OmegaConfGrammarParser.g4>`_
+for details):
+
+    - Unquoted strings (that support only a subset of characters, contrary to quoted ones): ``foo``, ``foo_bar``, ``hello world 123``
+    - Integer numbers: ``123``, ``-5``, ``+1_000_000``
+    - Floating point numbers (with special case-independent keywords for infinity and NaN): ``0.1``, ``1e-3``, ``inf``, ``-INF``, ``nan``
+    - Other special keywords (also case-independent): ``null``, ``true``, ``false``, ``NULL``, ``True``, ``fAlSe``.
+      **IMPORTANT**: ``None`` is *not* a special keyword and will be parsed as an unquoted string, you must
+      use the ``null`` keyword instead (as in YAML).
+    - Interpolations (thus allowing for nested interpolations)
+
+
+Escaped characters
+^^^^^^^^^^^^^^^^^^
+
+Some characters need to be escaped, with varying escaping requirements depending on the situation.
+In general, however, you can use the following rule of thumb:
+*you only need to escape characters that otherwise have a special meaning in the current context*.
+
+.. _escaping-in-interpolation-strings:
+
+Escaping in interpolation strings
++++++++++++++++++++++++++++++++++
+
+In order to define fields whose value is an interpolation-like string, interpolations can be escaped with ``\${``.
+For instance:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"\${dir}", "dir": "tmp"})
+    >>> print(c.path)  # does *not* interpolate into the `dir` node
+    ${dir}
+
+Since the ``\`` character has a special meaning in front of ``${``,
+it must itself be escaped with ``\\`` if you actually intend to follow it with an interpolation:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"C:\\${dir}", "dir": "tmp"})
+    >>> print(c.path)  # *does* interpolate into the `dir` node
+    C:\tmp
+
+Finally, since the ``\`` character has no special meaning unless followed by ``${``,
+it does *not* need to be escaped anywhere else:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"C:\foo_${dir}", "dir": "tmp"})
+    >>> print(c.path)  # a single \ is preserved...
+    C:\foo_tmp
+    >>> c = OmegaConf.create({"path": r"C:\\foo_${dir}", "dir": "tmp"})
+    >>> print(c.path)  # ... and multiple \\ too (no escape sequence)
+    C:\\foo_tmp
+
+Escaping in unquoted strings
+++++++++++++++++++++++++++++
+
+Unquoted strings can be found in a number of contexts, including dictionary keys/values,
+list elements, etc. As a result, the  escape sequences are used for some
+special characters
+(``\\``, ``\[``, ``\]``, ``\{``, ``\}``, ``\(``, ``\)``, ``\:``, ``\=``, ``\,``),
+for instance:
+
+    - ``C\:\\$\{dir\}`` resolves to the string ``"C:\${dir}"``
```

Comment:
Why do you have to escape the colon?

Reply:
Ah, good question, I forgot that escaping the colon is only required in dictionary keys for obvious reasons (but escaping is optional in other unquoted strings, i.e., both `\:` and `:` lead to the same result).

I'm tempted to not go into this detail for the documentation, because it's not a big deal if people escape a colon that doesn't require it, and in most situations I expect users to just use `:` without second thoughts and it will usually work.

Let me know though if you believe this should be fully documented here.